File: geo3x3.pas
{$mode objfpc}
unit geo3x3;

interface

type
  latlnglevelunit = array of real;

function Geo3x3_encode(lat: real; lng: real; level: integer): string;
function Geo3x3_decode(code: string): latlnglevelunit; static;

implementation

uses
  sysutils,
  math;


function Geo3x3_encode(lat: real; lng: real; level: integer): string;
var
  unitsize: real;
  i: integer;
  x: integer;
  y: integer;
begin
	if level < 1 then
		result := ''
  else
  begin
    result := 'E';
    if lng < 0.0 then
    begin
      result := 'W';
      lng := lng + 180.0;
    end;
		lat := lat + 90.0; { 180:the North Pole,  0:the South Pole }
		unitsize := 180.0;
		for i := 1 to level - 1 do
    begin
    	unitsize := unitsize / 3.0;
			x := Floor(lng / unitsize);
			y := Floor(lat / unitsize);
			result := result + IntToStr(x + y * 3 + 1);
			lng := lng - x * unitsize;
			lat := lat - y * unitsize;
		end
  end
end;

function Geo3x3_decode(code: string): latlnglevelunit; static;
var
  flg: boolean;
  unitsize: real;
  lat: real;
  lng: real;
  level: integer;
  n: integer;
  i: integer;
begin
  if length(code) = 0 then
    result := [0, 0, 0, 0]
  else
  begin
    flg := integer(code[1]) = 87; { ascii code of "W" }
		unitsize := 180.0;
		lat := 0.0;
		lng := 0.0;
		level := 1;
		for i := 1 to length(code) - 1 do
    begin
			n := integer(code[i + 1]) - 49; { ascii code of "1" }
      if (n < 0) or (n > 9) then
        break;
			unitsize := unitsize / 3.0;
			lng := lng + (n mod 3) * unitsize;
			lat := lat + Floor(n / 3) * unitsize;
			level := level + 1;
    end;
		lat := lat + unitsize / 2.0;
		lng := lng + unitsize / 2.0;
		lat := lat - 90.0;
    if flg then
      lng := lng - 180.0;
    result := [ lat, lng, level, unitsize];
  end;
end;

end.

File: geo3x3.f90
module geo3x3
  implicit none
contains
  character(50) function geo3x3_encode(lat, lng, level)
    real lat
    real lng
    integer level

    real flat
    real flng
    real unit
    integer i, x, y

    flat = lat
    flng = lng
    if (lng >= 0.0) then
      geo3x3_encode = "E"
    else
      geo3x3_encode = "W"
      flng = flng + 180.0
    end if
    flat = flat + 90.0
    unit = 180.0
    do i = 1, level - 1
      unit = unit / 3
      x = floor(flng / unit)
      y = floor(flat / unit)
      geo3x3_encode(i + 1: i + 1) = char(x + y * 3 + 1 + 48)
      flng = flng - x * unit
      flat = flat - y * unit
    end do
    !geo3x3_encode = "W"
  end

  function geo3x3_decode(code)
    character(len=*) code
    real geo3x3_decode(4)

    real lat
    real lng
    real level
    real unit
    integer flg, i, n

    lat = 0.0
    lng = 0.0
    level = 0
    unit = 180.0

    if (len(code) > 0) then
      flg = 0
      if (code(1:1) == "W") then
        flg = 1
      end if
      level = level + 1
      do i = 2, len(code)
        n = index("123456789", code(i:i)) - 1
        if (n >= 0) then
          unit = unit / 3
          lng = lng + mod(n, 3) * unit
          lat = lat + n / 3 * unit
          level = level + 1
        end if
      end do

      lat = lat + unit / 2
      lng = lng + unit / 2
      lat = lat - 90
      if (flg == 1) then
        lng = lng - 180.0
      end if
    end if
    geo3x3_decode = [lat, lng, level, unit]
  end
end
